### PR TITLE
Fixes #1091 issue with distributed parameters in individual bb

### LINKED ISSUE #1091: Issue with Distributed Parameters in Individual BB
Right now, the concept has a serious flow (I think)
Consider the Kidney parameter, volume

image

It's a distributed parameter, you can't see the value. Instead, there are TWO other parameters that are somewhat irrelevant for the standard user

image

How should we change the display?

@Yuri05 @rwmcintosh @PavelBal @stephanschaller @georgeDaskalakis

### PR DESCRIPTION
**Pull request by rwmcintosh**
Fixes #1091

# Description
In the Individual Building Block view, you would see distribution parameters separate from their distributed parameter container. This introduces a new type of editor for this view that shows the distribution parameters in a combo box popup.

## Type of change

Please mark relevant options with an `x` in the brackets.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires documentation changes (link at least one [user](https://github.com/Open-Systems-Pharmacology/docs) or [developer](https://github.com/Open-Systems-Pharmacology/developer-docs) documentation issue):
- [ ] Algorithm update - updates algorithm documentation/questions/answers etc.
- [ ] Other (please describe):

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Integration tests
- [x] Unit tests
- [x] Manual tests 
- [ ] No tests required

# Reviewer checklist

Mark everything that needs to be checked before merging the PR.

- [ ] Check if the code is well documented
- [ ] Check if the behavior is what is expected
- [ ] Check if the code is well tested
- [ ] Check if the code is readable and well formatted
- [ ] Additional checks (document below if any)
- [ ] Check if documentation update issue(s) are created if the option `This change requires a documentation update` above is selected

# Screenshots (if appropriate):
image

image

The parameters included depend on the distribution type.

# Questions (if appropriate):

### CODE REVIEW

File: src/MoBi.Presentation/Presenter/PathAndValueBuildingBlockPresenter.cs [hunk 1]
```diff
@@ -1,303 +1,99 @@
-﻿using System;
-using System.Collections.Generic;
-using System.ComponentModel;
+﻿using System.Collections.Generic;
```
Comment (rwmcintosh) [inline src/MoBi.Presentation/Presenter/PathAndValueBuildingBlockPresenter.cs:1]: There was a file level renaming done with these presenters. One was called PathWithValue... which doesn't match anything so this looks like a lot of changes, but really is not.

File: src/MoBi.Presentation/Presenter/PathAndValueModuleBuildingBlockPresenter.cs [hunk 2]
```diff
@@ -0,0 +1,303 @@
+﻿using System;
```
Comment (rwmcintosh) [inline src/MoBi.Presentation/Presenter/PathAndValueModuleBuildingBlockPresenter.cs:1]: This file is not new, just part of a rename gone wrong

File: src/MoBi.UI/Views/ExpressionProfileBuildingBlockView.cs [hunk 3]
```diff
@@ -33,12 +33,26 @@ public partial class ExpressionProfileBuildingBlockView : BaseUserControl, IExpr
...
+         _repositoryItemPopupContainerEdit.PopupControl = _popupControl;
+         _repositoryItemPopupContainerEdit.QueryDisplayText += (o, e) => OnEvent(queryText, e);
+      }
+
+      private void queryText(QueryDisplayTextEventArgs e)
+      {
+         var distributedParameter = _gridViewBinder.FocusedElement;
+         if (distributedParameter == null)
+            return;
+         e.DisplayText = distributedParameter.ExpressionParameterFormatter().Format(distributedParameter.Value);
```
Comment (rwmcintosh) [inline src/MoBi.UI/Views/ExpressionProfileBuildingBlockView.cs:55]: This is the line responsible for the formatting when the popup is being shown. With this line, it does not change when it's focused, unfocused, or when the popup is showing.

File: src/MoBi.Presentation/DTO/PathAndValueEntityDTO.cs [hunk 4]
```diff
@@ -1,16 +1,55 @@
...
+      }
+
+      public void AddSubParameter(TSubParameter subParameter)
+      {
+         _subParameters.Add(subParameter);
+      }
+
+      public bool IsDistributed => PathWithValueObject.DistributionType != null;
+
+      public string DistributionType => $"{PathWithValueObject.DistributionType} {Distribution}";
```
Comment (msevestre) [inline src/MoBi.Presentation/DTO/PathAndValueEntityDTO.cs:36]: THis will look weird if it's not Distributed
PathWithValueObject.DistributionType==null

Comment (msevestre) [reply src/MoBi.Presentation/DTO/PathAndValueEntityDTO.cs:36]: I don't understand what the text is going to be. What is Distributiion in this case?

File: src/MoBi.Presentation/DTO/PathAndValueEntityDTO.cs [hunk 5]
```diff
@@ -1,16 +1,55 @@
...
+
+      public override Unit DisplayUnit
+      {
+         get => IsDistributed ? MeanDTO?.DisplayUnit : base.DisplayUnit;
+         set => base.DisplayUnit = value;
```
Comment (msevestre) [inline src/MoBi.Presentation/DTO/PathAndValueEntityDTO.cs:41]: can this be called for Distrbitued paramter?

File: src/MoBi.Presentation/Mappers/PathAndValueEntityBuildingBlockToPathAndValueEntityBuildingBlockDTOMapper.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+      private readonly IMapper<TBuilder, TBuilderDTO> _builderToBuilderDTOMapper;
+
+      protected PathAndValueEntityBuildingBlockToPathAndValueEntityBuildingBlockDTOMapper(IMapper<TBuilder, TBuilderDTO> builderToBuilderDTOMapper)
+      {
+         _builderToBuilderDTOMapper = builderToBuilderDTOMapper;
+      }
+
+      public TBuildingBlockDTO MapFrom(TBuildingBlock buildingBlock)
+      {
+         var parameterDTOs = buildingBlock.MapAllUsing(_builderToBuilderDTOMapper).ToList();
```
Comment (msevestre) [inline src/MoBi.Presentation/Mappers/PathAndValueEntityBuildingBlockToPathAndValueEntityBuildingBlockDTOMapper.cs:25]: this already returns an IReadonLy. No need to ToList again

File: src/MoBi.UI/Views/ExpressionProfileBuildingBlockView.cs (hunk 3 above)
Comment (msevestre) [inline src/MoBi.UI/Views/ExpressionProfileBuildingBlockView.cs:55]: That's a  weird extensuion method. Why is the formatter not a property of the view?

File: src/MoBi.Presentation/DTO/PathAndValueEntityDTO.cs (hunk 4 above)
Comment (msevestre) [reply src/MoBi.Presentation/DTO/PathAndValueEntityDTO.cs:36]: oh I see this is the text from constant "Distribution"

**Review by msevestre — APPROVED** [review]
Looks all very good. Just suprised about this ugly extension methods to return a formatted.

**Comment by msevestre** [comment]
But i guess it's not new :)

**Pull request merged**